Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,308 @@
+    veblenWith f a b ≤ veblenWith f c d ↔
+    a = c ∧ b ≤ d ∨ a < c ∧ b ≤ veblenWith f c d ∨ c < a ∧ veblenWith f a b ≤ d := by
+  obtain h | rfl | h := lt_trichotomy a c
+  · simp_rw [h, h.ne, h.not_lt, false_and, false_or, or_false, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_le_veblenWith_right_iff hf]
+  · simp [veblenWith_le_veblenWith_right_iff hf]
+  · simp_rw [h, h.ne', h.not_lt, false_and, false_or, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_le_veblenWith_right_iff hf]
+
+theorem veblenWith_eq_veblenWith_iff (hf : IsNormal f) {a b c d : Ordinal} :
+    veblenWith f a b = veblenWith f c d ↔
+    a = c ∧ b = d ∨ a < c ∧ b = veblenWith f c d ∨ c < a ∧ veblenWith f a b = d := by
+  obtain h | rfl | h := lt_trichotomy a c
+  · simp_rw [h, h.ne, h.not_lt, false_and, false_or, or_false, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_inj hf]
+  · simp [veblenWith_inj hf]
+  · simp_rw [h, h.ne', h.not_lt, false_and, false_or, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_inj hf]
+
+/-! ### Veblen function -/
+
+private theorem isNormal_omega0_opow : IsNormal fun a ↦ ω ^ a := isNormal_opow one_lt_omega0
+private theorem omega0_opow_zero_pos : 0 < ω ^ (0 : Ordinal) := by simp
+
+/-- `veblen o` is the `o`-th function in the Veblen hierarchy starting with `ω ^ ·`. That is:
+
+- `veblen 0 a = ω ^ a`.
+- `veblen a` enumerates the fixed points of `veblen b` for `b < a` when `a ≠ 0`.
+-/
+@[pp_nodot]
+def veblen : Ordinal.{u} → Ordinal.{u} → Ordinal.{u} :=
+  veblenWith (ω ^ ·)
+
+@[simp]
+theorem veblen_zero : veblen 0 = fun a ↦ ω ^ a := by
+  rw [veblen, veblenWith_zero]
+
+theorem veblen_zero_apply (o : Ordinal) : veblen 0 o = ω ^ o := by
+  rw [veblen_zero]
+
+theorem veblen_of_ne_zero (h : o ≠ 0) : veblen o = derivFamily fun x : Set.Iio o ↦ veblen x.1 :=
+  veblenWith_of_ne_zero _ h
+
+theorem isNormal_veblen (o : Ordinal) : IsNormal (veblen o) :=
+  isNormal_veblenWith isNormal_omega0_opow o
+
+theorem veblen_veblen_of_lt (h : a < b) (c : Ordinal) : veblen a (veblen b c) = veblen b c :=
+  veblenWith_veblenWith_of_lt isNormal_omega0_opow h c
+
+theorem veblen_succ (o : Ordinal) : veblen (Order.succ o) = deriv (veblen o) :=
+  veblenWith_succ isNormal_omega0_opow o
+
+theorem veblen_right_strictMono (o : Ordinal) : StrictMono (veblen o) :=
+  veblenWith_right_strictMono isNormal_omega0_opow o
+
+theorem veblen_right_monotone (o : Ordinal) : Monotone (veblen o) :=
+  (veblen_right_strictMono o).monotone
+
+@[simp]
+theorem veblen_lt_veblen_right_iff : veblen o a < veblen o b ↔ a < b :=
+  veblenWith_lt_veblenWith_right_iff isNormal_omega0_opow
+
+@[simp]
+theorem veblen_le_veblen_right_iff : veblen o a ≤ veblen o b ↔ a ≤ b :=
+  veblenWith_le_veblenWith_right_iff isNormal_omega0_opow
+
+@[simp]
+theorem veblen_inj : veblen o a = veblen o b ↔ a = b :=
+  veblenWith_inj isNormal_omega0_opow
```

Comment:
Could you add an `injective` version of this, with conclusion `Function.Injective (veblen o)`?